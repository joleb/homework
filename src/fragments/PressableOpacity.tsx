import React from "react";
import {
  Pressable,
  PressableProps,
  PressableStateCallbackType,
  StyleSheet,
  View,
} from "react-native";

const defaultHitSlop = { top: 8, right: 8, bottom: 8, left: 8 };

const PressableOpacity: React.FC<
  PressableProps & {
    children: React.ReactNode;
  }
> = ({ children, style, hitSlop = defaultHitSlop, ...props }) => {
  const customStyle: PressableProps["style"] = (
    state: PressableStateCallbackType,
  ) =>
    StyleSheet.compose(
      {
        opacity: state.pressed ? 0.5 : 1,
      },
      // use style  if provided and take into account the state
      style && typeof style === "function" ? style(state) : style,
    );
  return (
    <Pressable hitSlop={hitSlop} style={customStyle} {...props}>
      <View pointerEvents="box-none">{children}</View>
    </Pressable>
  );
};

export default PressableOpacity;
