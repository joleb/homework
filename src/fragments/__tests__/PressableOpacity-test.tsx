import React from "react";
import { Text } from "react-native";

import { render, fireEvent } from "@testing-library/react-native";

import PressableOpacity from "../PressableOpacity";

describe("PressableOpacity", () => {
  it("matches snapshot", () => {
    const { toJSON } = render(
      <PressableOpacity onPress={() => {}}>
        <Text>Press me</Text>
      </PressableOpacity>,
    );
    expect(toJSON()).toMatchSnapshot();
  });
  it("renders children correctly", () => {
    const { getByText } = render(
      <PressableOpacity onPress={() => {}}>
        <Text>Press me</Text>
      </PressableOpacity>,
    );

    expect(getByText("Press me")).toBeTruthy();
  });

  it("calls onPress when pressed", () => {
    const onPressMock = jest.fn();
    const { getByText } = render(
      <PressableOpacity onPress={onPressMock}>
        <Text>Press me</Text>
      </PressableOpacity>,
    );

    fireEvent.press(getByText("Press me"));
    expect(onPressMock).toHaveBeenCalled();
  });

  it("changes opacity when pressed", () => {
    const { getByTestId } = render(
      <PressableOpacity onPress={() => {}} testID="pressable" testOnly_pressed>
        <Text>Press me</Text>
      </PressableOpacity>,
    );

    const pressable = getByTestId("pressable");
    fireEvent.press(pressable);
    expect(pressable?.props?.style).toMatchObject({ opacity: 0.5 });
  });
});
