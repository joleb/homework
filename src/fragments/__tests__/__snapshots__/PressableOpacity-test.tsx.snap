// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PressableOpacity matches snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  hitSlop={
    {
      "bottom": 8,
      "left": 8,
      "right": 8,
      "top": 8,
    }
  }
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
>
  <View
    pointerEvents="box-none"
  >
    <Text>
      Press me
    </Text>
  </View>
</View>
`;
